content/browser/ml/webnn/dml/readback_heap.h

```diff
@@ -0,0 +1,53 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef CONTENT_BROWSER_ML_WEBNN_READBACK_HEAP_H_
+#define CONTENT_BROWSER_ML_WEBNN_READBACK_HEAP_H_
+
+#include <wrl.h>
+#include <vector>
+
+#include "DirectML.h"
+#include "components/ml/mojom/webnn_graph.mojom.h"
+#include "content/browser/ml/webnn/dml/utils_dml.h"
+
+namespace content::webnn {
+
+using Microsoft::WRL::ComPtr;
+using ml::webnn::mojom::NamedOutputsPtr;
+
+class ExecutionContext;
+
+class ReadbackHeap final {
```

Comment:
This class is a bit confusing because none of this code creates a "read-back heap" in the sense how DML defines a heap. This class appears to be uploading CPU data from a shared buffer to a DML resource, which is "uploader" would be more appropriate.